### PR TITLE
Implement histogram smoothing

### PR DESCRIPTION
**Pull request by mark-raymond**

### CODE REVIEW

File: Sources/Extraction/Filters/LocalHistogram.c [hunk 1]
```diff
@@ -18,20 +18,23 @@ Int16Array3D LocalHistogram_Analyze(const BlockMap *blocks, const UInt8Array2D *
     return histogram;
 }
 
-Int16Array3D LocalHistogram_SmoothAroundCorners(const BlockMap *blocks, const Int16Array3D *input)
+Int16Array3D LocalHistogram_SmoothAroundCorners(const Int16Array3D *input)
 {
-    Int16Array3D output;
+    Int16Array3D output = Int16Array3D_Construct(input->sizeX + 1, input->sizeY + 1, input->sizeZ);
```
Comment (ghost) [inline Sources/Extraction/Filters/LocalHistogram.c]: I would say we don't add +1 here as clients may not expect this to be bigger than the input image size

Comment (ghost) [inline Sources/Extraction/Filters/LocalHistogram.c]: Also a pattern we're using is to take in a pre-allocated pointer for the output array.

i.e. the function signature would be
`void LocalHistogram_SmoothAroundCorners(const Int16Array3D *input, Int16Array3D *output)`

This way the client knows it's their responsibility to allocate the memory and subsequently free it.

**Comment by ghost** [comment]
A couple of minor points but otherwise looks great :)

File: Sources/Extraction/Filters/LocalHistogram.c (hunk 1 above)
Comment (mark-raymond) [inline Sources/Extraction/Filters/LocalHistogram.c]: Re the +1: that's what the C# code does, so we copied it.

**Comment by mark-raymond** [comment]
We've changed the signature round so the caller allocates & frees.

File: Sources/Extraction/Filters/LocalHistogram.c [hunk 2]
```diff
@@ -2,36 +2,34 @@
...
 
-Int16Array3D LocalHistogram_Smooth(const BlockMap *blocks, const Int16Array3D *input)
+void LocalHistogram_SmoothAroundCorners(const Int16Array3D *input, Int16Array3D *output)
 {
-    Int16Array3D output;
-
-    /* TODO: Implement LocalHistogram_Smooth() */
-
-    return output;
+    for (int x = 0; x < input->sizeX; x++)
```
Comment (ghost) [inline Sources/Extraction/Filters/LocalHistogram.c:26]: We should probably assert that output is the size we expect (the same size as input) and also isn't null

File: Sources/Extraction/Filters/LocalHistogram.c [hunk 3]
```diff
@@ -2,36 +2,34 @@
...
+    {
+        for (int y = 0; y < input->sizeY; y++)
+        {
+            for (int z = 0; z < input->sizeZ; z++)
+            {
+                output->data[x][y][z] += input->data[x][y][z];
+                output->data[x][y+1][z] += input->data[x][y][z];
```
Comment (ghost) [inline Sources/Extraction/Filters/LocalHistogram.c:33]: We'll need to check for out of range access now as output should be the same size as input

File: Sources/Extraction/Filters/LocalHistogram.c (hunk 2 above)
Comment (breggles) [inline Sources/Extraction/Filters/LocalHistogram.c:26]: So, as mentioned in our previous reply, SourceAFIS used +1 corners. What would be the reason to deviate from SourceAFIS, on this?

Comment (ghost) [inline Sources/Extraction/Filters/LocalHistogram.c:26]: Oh yes, you're right! I didn't see that previous reply - it disappeared from Github. Sorry.

Comment (ghost) [inline Sources/Extraction/Filters/LocalHistogram.c:26]: An

`assert(output.sizeX == input.sizeX+1);` and the same for sizeY might be quite useful though

File: Sources/Tests/Extraction/Filters/Test_LocalHistogram.c [hunk 4]
```diff
@@ -28,9 +28,13 @@ TEST(LocalHistogram, LocalHistogram_Analyze_same_values_go_into_same_bucket)
     image.data[0][0] = 3;
     image.data[0][1] = 3;
 
-    Int16Array3D histogram = LocalHistogram_Analyze(&blocks, &image);
+    Int16Array3D histogram = Int16Array3D_Construct(blocks.blockCount.width, blocks.blockCount.height, 256);
```
Comment (ghost) [inline Sources/Tests/Extraction/Filters/Test_LocalHistogram.c:31]: Isn't this cornerCount.width + 1 and cornerCount.height+1?

**Pull request merged**